src/HashStore.hs

```diff
@@ -1,16 +1,49 @@
 module HashStore
-       ( hashFile
+       ( hashStore
        ) where
 
 import Crypto.Hash.BLAKE2.BLAKE2b (hash)
-import Data.ByteString.Char8 (ByteString, unpack)
+import Data.ByteString.Char8 (ByteString)
+import System.Directory (createDirectoryIfMissing, doesFileExist)
+import System.FilePath ((</>))
 
+import qualified Data.ByteString.Char8 as BS
+
+-- QUESTION: does it make sense to create `newtype Hash = Hash ByteString`?
 -- | Perform 'ByteString' hashing.
 hashContent :: ByteString -> ByteString
 hashContent = hash 64 mempty
 {-# INLINE hashContent #-}
 
 -- | Build the new 'FilePath' adding the hash of the provided content.
 hashFile :: ByteString -> FilePath -> FilePath
-hashFile content file = unpack (hashContent content) ++ "-" ++ file
+hashFile content file = BS.unpack (hashContent content) ++ "-" ++ file
 {-# INLINE hashFile #-}
+
+{- | @'hashStore' storePath action (name, content)@ computes @hash@ of @content@
+and performs @action@ if file @storePath '</>' (hash ++ "-" ++ name)@ doesn't
+exist, writing new content in this file.
+
+The property is that for any item @(name, hash)@, we get a valid filepath by
+@storePath '</>' (hash ++ "-" ++ name)@.
+-}
+hashStore :: FilePath
+             -- ^ Directory to store file contents with new hash
+          -> (ByteString -> IO ByteString)
+             -- ^ Action to be performed if hash is different
+          -> (String, ByteString)
+             -- ^ File name
+          -> IO (String, ByteString)
+             -- ^ Hash of given 'ByteString' and new 'ByteString' if old hash and new one are different
+hashStore storePath action (name, content) = do
+    createDirectoryIfMissing True storePath
+    let hashName = hashFile content name
+    let storeName = storePath </> hashName
+
+    isUpToDate <- doesFileExist storeName
+    if isUpToDate then
+        return (hashName, content)
+    else do
+        newContent <- action content
```

Comment:
I've read the comment on line 37 and now I realize this could be intentional, not s bug. Well, in this case the output bytestring becomes completely uselss: if tge caller doesn't know whether it's the input or the output, what are they supposed to do with it? It should always be the output.

Reply:
In terms of efficiency, there are actually two use cases:

1. The caller needs the file content. Then it's better to return the bytestring (we avoid the overhead of reading the file in the cache miss case)

2. The caller doesn't need the content. Then it is better not to return the bytestring (we avoid the overhrad of reading the file in the cache hit case).

I propose to have two functions then: hashStore and hashStoreWithContent